prov/verbs/src/fi_verbs.h

```diff
@@ -394,33 +443,177 @@ void fi_ibv_monitor_unsubscribe(struct ofi_mem_monitor *notifier,
+	struct sockaddr			*peer_addr;
+	struct fi_ibv_cq		*tx_cq;
+
+	/* The physical INI/TGT QPN connection. Virtual connections to the
+	 * same remote peer and TGT QPN will share this connection, with
+	 * the remote end opening the specified XRC TGT QPN for sharing. */
+	enum fi_ibv_ini_qp_state	state;
+	struct ibv_qp			*ini_qp;
+	uint32_t			tgt_qpn;
+
+	/* EP waiting on or using this INI/TGT physical connection will be in
+	 * one of these list and hold a reference to the shared connection. */
+	struct dlist_entry		pending_list;
+	struct dlist_entry		active_list;
+	ofi_atomic32_t			ref_cnt;
+};
+
+enum fi_ibv_xrc_ep_conn_state {
+	FI_IBV_XRC_UNCONNECTED,
+	FI_IBV_XRC_ORIG_CONNECTING,
+	FI_IBV_XRC_ORIG_CONNECTED,
+	FI_IBV_XRC_RECIP_CONNECTING,
+	FI_IBV_XRC_CONNECTED
+};
+
+/*
+ * The following XRC state is only required during XRC connection
+ * establishment and can be freed once bidirectional connectivity
+ * is established.
+ */
+struct fi_ibv_xrc_ep_conn_setup {
+	uint32_t			tag;
+
+	/* IB CM message stale/duplicate detection processing requires
+	 * that shared INI/TGT connections use unique QP numbers during
+	 * RDMA CM connection setup. To avoid conflicts with actual HCA
+	 * QP number space, we allocate minimal QP that are left in the
+	 * reset state and closed once the setup process completes. */
+	struct ibv_qp			*rsvd_ini_qpn;
+	struct ibv_qp			*rsvd_tgt_qpn;
+
+
+	/* Delivery of the FI_CONNECTED event is delayed until
+	 * bidirectional connectivity is established. */
+	size_t				event_len;
+	uint8_t				event_data[FI_IBV_CM_DATA_SIZE];
+
+	/* Connection request may have to queue waiting for the
+	 * physical XRC INI/TGT QP connection to complete. */
+	int				pending_recip;
+	size_t				pending_paramlen;
+	uint8_t				pending_param[FI_IBV_CM_DATA_SIZE];
+};
+
 struct fi_ibv_ep {
-	struct util_ep		util_ep;
-	struct ibv_qp		*ibv_qp;
+	struct util_ep			util_ep;
+	struct ibv_qp			*ibv_qp;
 	union {
-		struct rdma_cm_id	*id;
+		struct rdma_cm_id		*id;
 		struct {
 			struct ofi_ib_ud_ep_name	ep_name;
 			int				service;
 		};
 	};
-	struct fi_ibv_eq	*eq;
-	struct fi_ibv_srq_ep	*srq_ep;
-	struct fi_info		*info;
+	struct fi_ibv_eq		*eq;
+	struct fi_ibv_srq_ep		*srq_ep;
+	struct fi_info			*info;
 
 	struct {
 		struct ibv_send_wr	rma_wr;
 		struct ibv_send_wr	msg_wr;
-		struct ibv_sge		sge;	
+		struct ibv_sge		sge;
 	} *wrs;
+
+#ifdef INCLUDE_VERBS_XRC
+	/* XRC only fields */
+	struct rdma_cm_id		*tgt_id;
+	struct ibv_qp			*tgt_ibv_qp;
+	enum fi_ibv_xrc_ep_conn_state	conn_state;
+	struct fi_info			*tgt_info;
+	uint32_t			srqn;
+	uint32_t			peer_srqn;
+
+	/* A reference is held to a shared physical XRC INI/TGT QP connecting
+	 * to the destination node. */
+	struct fi_ibv_ini_shared_conn	*ini_conn;
+	struct dlist_entry		ini_conn_entry;
+
+	/* The following state is allocated during XRC bidirectional setup and
+	 * freed once the connection is established. */
+	struct fi_ibv_xrc_ep_conn_setup	*conn_setup;
+#endif /* INCLUDE_VERBS_XRC */
 };
```

Comment:
Please create an xrc specific ep structure, rather than extending the base structure.  Also, this seems like a lot of per ep overhead.

Reply:
This is done now. The RC endpoint is not impacted by inclusion of XRC and the XRC endpoint is aprox 50 bytes larger. Note that the XRC endpoint that shares the XRC QP connection does not create another IB QP TX work queue, so this is a significant memory savings.